In [145]:
# Import Required Libraries
import pandas as pd
import numpy as np

In [146]:
# pandas options
pd.set_option("display.max.columns", None)
pd.set_option("display.max.rows", None)
pd.set_option("display.precision", 2)

In [147]:
# Load the Hip Replacement CCG data
file_path = r"c:\Users\mittall\source\EAISI\NHS\EAISI-NHS\notebooks\Lalita\UnderstandingData\merged\Hip Replacement Provider.csv"
df = pd.read_csv(file_path, sep=';')

# Make a copy of the original dataframe
df_original = df.copy()
print("Original dataframe copied to df_original.")

Original dataframe copied to df_original.


C:\Users\mittall\AppData\Local\Temp\ipykernel_19812\1061808804.py:3: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path, sep=';')


In [148]:
# Display basic information about the dataset
print("Shape of the dataset:", df.shape)
print("\nColumn names:")
print(df.columns.tolist())
print("\nData types:")
print(df.dtypes)
print("\nFirst 5 rows:")
df.head()

Shape of the dataset: (124844, 82)

Column names:
['Provider Code', 'Procedure', 'Revision Flag', 'Year', 'Age Band', 'Gender', 'Pre-Op Q Assisted', 'Pre-Op Q Assisted By', 'Pre-Op Q Symptom Period', 'Pre-Op Q Previous Surgery', 'Pre-Op Q Living Arrangements', 'Pre-Op Q Disability', 'Heart Disease', 'High Bp', 'Stroke', 'Circulation', 'Lung Disease', 'Diabetes', 'Kidney Disease', 'Nervous System', 'Liver Disease', 'Cancer', 'Depression', 'Arthritis', 'Pre-Op Q Mobility', 'Pre-Op Q Self-Care', 'Pre-Op Q Activity', 'Pre-Op Q Discomfort', 'Pre-Op Q Anxiety', 'Pre-Op Q EQ5D Index Profile', 'Pre-Op Q EQ5D Index', 'Post-Op Q Assisted', 'Post-Op Q Assisted By', 'Post-Op Q Living Arrangements', 'Post-Op Q Disability', 'Post-Op Q Mobility', 'Post-Op Q Self-Care', 'Post-Op Q Activity', 'Post-Op Q Discomfort', 'Post-Op Q Anxiety', 'Post-Op Q Satisfaction', 'Post-Op Q Sucess', 'Post-Op Q Allergy', 'Post-Op Q Bleeding', 'Post-Op Q Wound', 'Post-Op Q Urine', 'Post-Op Q Further Surgery', 'Post-Op Q R

,Provider Code,Procedure,Revision Flag,Year,Age Band,Gender,Pre-Op Q Assisted,Pre-Op Q Assisted By,Pre-Op Q Symptom Period,Pre-Op Q Previous Surgery,Pre-Op Q Living Arrangements,Pre-Op Q Disability,Heart Disease,High Bp,Stroke,Circulation,Lung Disease,Diabetes,Kidney Disease,Nervous System,Liver Disease,Cancer,Depression,Arthritis,Pre-Op Q Mobility,Pre-Op Q Self-Care,Pre-Op Q Activity,Pre-Op Q Discomfort,Pre-Op Q Anxiety,Pre-Op Q EQ5D Index Profile,Pre-Op Q EQ5D Index,Post-Op Q Assisted,Post-Op Q Assisted By,Post-Op Q Living Arrangements,Post-Op Q Disability,Post-Op Q Mobility,Post-Op Q Self-Care,Post-Op Q Activity,Post-Op Q Discomfort,Post-Op Q Anxiety,Post-Op Q Satisfaction,Post-Op Q Sucess,Post-Op Q Allergy,Post-Op Q Bleeding,Post-Op Q Wound,Post-Op Q Urine,Post-Op Q Further Surgery,Post-Op Q Readmitted,Post-Op Q EQ5D Index Profile,Post-Op Q EQ5D Index,Hip Replacement EQ5D Index Post-Op Q Predicted,Pre-Op Q EQ VAS,Post-Op Q EQ VAS,Hip Replacement EQ VAS Post-Op Q Predicted,Hip Replacement Pre-Op Q Pain,Hip Replacement Pre-Op Q Sudden Pain,Hip Replacement Pre-Op Q Night Pain,Hip Replacement Pre-Op Q Washing,Hip Replacement Pre-Op Q Transport,Hip Replacement Pre-Op Q Dressing,Hip Replacement Pre-Op Q Shopping,Hip Replacement Pre-Op Q Walking,Hip Replacement Pre-Op Q Limping,Hip Replacement Pre-Op Q Stairs,Hip Replacement Pre-Op Q Standing,Hip Replacement Pre-Op Q Work,Hip Replacement Pre-Op Q Score,Hip Replacement Post-Op Q Pain,Hip Replacement Post-Op Q Sudden Pain,Hip Replacement Post-Op Q Night Pain,Hip Replacement Post-Op Q Washing,Hip Replacement Post-Op Q Transport,Hip Replacement Post-Op Q Dressing,Hip Replacement Post-Op Q Shopping,Hip Replacement Post-Op Q Walking,Hip Replacement Post-Op Q Limping,Hip Replacement Post-Op Q Stairs,Hip Replacement Post-Op Q Standing,Hip Replacement Post-Op Q Work,Hip Replacement Post-Op Q Score,Hip Replacement OHS Post-Op Q Predicted,CSVYear
0,ADP02,Hip Replacement,0,2016/17,*,*,2,0,2,1,1,1,9,9,9,9,9,9,1,9,9,9,9,1,9,9,9,9,9,99999,NaN,2,9,2,2,1,1,2,1,1,2,1,2,2,2,2,2,2,11211,0.88,NaN,999,80,NaN,0,2,1,3,1,2,3,2,1,0,2,2,19.0,3,4,2,4,3,2,4,4,3,2,2,3,36.0,3.777.303.196,2016/17
1,ADP02,Hip Replacement,0,2016/17,*,*,2,0,3,2,1,1,9,9,9,9,9,9,9,9,9,9,9,9,2,2,2,3,1,22231,0.06,2,9,1,2,2,1,1,1,1,1,1,2,2,2,2,2,2,21111,0.85,0.67,30,70,6.592.244.892,0,1,0,2,2,1,1,1,0,2,2,1,13.0,4,4,4,4,3,3,4,4,4,4,4,4,46.0,3.558.681.573,2016/17
2,ADP02,Hip Replacement,0,2016/17,*,*,1,0,2,2,1,1,9,1,9,9,9,9,9,9,9,9,9,1,2,2,3,3,2,22332,-0.07,2,9,1,1,2,1,2,1,1,2,1,2,2,2,2,2,2,21211,0.81,0.70,40,60,6.810.332.786,1,0,0,1,1,1,0,1,0,1,1,0,7.0,1,4,3,3,2,3,2,4,1,2,4,2,31.0,3.293.405.433,2016/17
3,ADP02,Hip Replacement,0,2016/17,*,*,2,0,4,2,1,2,9,9,9,9,9,9,9,1,9,9,9,9,2,2,2,3,1,22231,0.06,2,9,1,2,2,1,2,2,2,3,2,2,2,2,1,2,1,21222,0.62,0.71,50,55,6.883.315.758,0,4,0,2,2,1,3,2,1,2,2,2,21.0,0,0,0,3,2,2,2,1,1,2,2,0,15.0,3.867.962.866,2016/17
4,ADP02,Hip Replacement,0,2016/17,*,*,2,0,4,2,1,2,9,9,9,9,9,9,9,9,9,9,9,9,2,2,2,3,2,22232,-0.02,2,9,1,2,1,1,1,1,1,1,1,2,2,2,2,2,2,11111,1.00,0.74,20,90,7.226.409.643,0,0,0,1,1,1,2,2,0,2,1,0,10.0,4,4,3,4,4,4,4,4,4,4,4,4,47.0,3.686.746.026,2016/17


In [149]:
# Handle missing values and data cleaning
# Replace common missing value indicators with NaN
df.replace({9: np.nan, '*': np.nan, '': np.nan, 999: np.nan}, inplace=True)

# Convert numeric columns to appropriate types
# First, identify columns that should be numeric
#identify the numerical columns
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
numerical_df = df[numerical_cols]
print("List of numerical columns:")
print(numerical_cols)
print(numerical_df.shape)
print(df.describe())





List of numerical columns:
['Revision Flag', 'Pre-Op Q Assisted', 'Pre-Op Q Assisted By', 'Pre-Op Q Symptom Period', 'Pre-Op Q Previous Surgery', 'Pre-Op Q Living Arrangements', 'Pre-Op Q Disability', 'Heart Disease', 'High Bp', 'Stroke', 'Circulation', 'Lung Disease', 'Diabetes', 'Kidney Disease', 'Nervous System', 'Liver Disease', 'Cancer', 'Depression', 'Arthritis', 'Pre-Op Q Mobility', 'Pre-Op Q Self-Care', 'Pre-Op Q Activity', 'Pre-Op Q Discomfort', 'Pre-Op Q Anxiety', 'Pre-Op Q EQ5D Index Profile', 'Pre-Op Q EQ5D Index', 'Post-Op Q Assisted', 'Post-Op Q Assisted By', 'Post-Op Q Living Arrangements', 'Post-Op Q Disability', 'Post-Op Q Mobility', 'Post-Op Q Self-Care', 'Post-Op Q Activity', 'Post-Op Q Discomfort', 'Post-Op Q Anxiety', 'Post-Op Q Satisfaction', 'Post-Op Q Sucess', 'Post-Op Q Allergy', 'Post-Op Q Bleeding', 'Post-Op Q Wound', 'Post-Op Q Urine', 'Post-Op Q Further Surgery', 'Post-Op Q Readmitted', 'Post-Op Q EQ5D Index Profile', 'Post-Op Q EQ5D Index', 'Hip Replacemen

In [150]:
# Identify numerical and string columns

string_cols = df.select_dtypes(include=['object']).columns.tolist()

# Create separate dataframes for numerical and string data

string_df = df[string_cols]

# Print summary information
print("DATASET OVERVIEW:")
print(f"Total observations: {len(df)}")
print(f"Total variables/columns: {len(df.columns)}")
print(f"\nNumerical variables/columns: {len(numerical_cols)}")
print(f"String variables/columns: {len(string_cols)}")

print(f"\nNumerical columns/columns ({len(numerical_cols)}):")
print(numerical_cols)

print(f"\nString columns/columns ({len(string_cols)}):")
print(string_cols)

print(f"\nNumerical dataframe shape: {numerical_df.shape}")
print(f"String dataframe shape: {string_df.shape}")

DATASET OVERVIEW:
Total observations: 124844
Total variables/columns: 82

Numerical variables/columns: 74
String variables/columns: 8

Numerical columns/columns (74):
['Revision Flag', 'Pre-Op Q Assisted', 'Pre-Op Q Assisted By', 'Pre-Op Q Symptom Period', 'Pre-Op Q Previous Surgery', 'Pre-Op Q Living Arrangements', 'Pre-Op Q Disability', 'Heart Disease', 'High Bp', 'Stroke', 'Circulation', 'Lung Disease', 'Diabetes', 'Kidney Disease', 'Nervous System', 'Liver Disease', 'Cancer', 'Depression', 'Arthritis', 'Pre-Op Q Mobility', 'Pre-Op Q Self-Care', 'Pre-Op Q Activity', 'Pre-Op Q Discomfort', 'Pre-Op Q Anxiety', 'Pre-Op Q EQ5D Index Profile', 'Pre-Op Q EQ5D Index', 'Post-Op Q Assisted', 'Post-Op Q Assisted By', 'Post-Op Q Living Arrangements', 'Post-Op Q Disability', 'Post-Op Q Mobility', 'Post-Op Q Self-Care', 'Post-Op Q Activity', 'Post-Op Q Discomfort', 'Post-Op Q Anxiety', 'Post-Op Q Satisfaction', 'Post-Op Q Sucess', 'Post-Op Q Allergy', 'Post-Op Q Bleeding', 'Post-Op Q Wound', 'Po

In [152]:
# Check for missing values
print("Missing values per column:")
print(df.isnull().sum())
print("\nPercentage of missing values:")
print((df.isnull().sum() / len(df)) * 100)

# Display first 10 rows
print("\nFirst 10 rows:")
df.head(10)

# Display last 10 rows
print("\nLast 10 rows:")
df.tail(10)

Missing values per column:
Provider Code                                          0
Procedure                                              0
Revision Flag                                          0
Year                                                   0
Age Band                                            9591
Gender                                              9591
Pre-Op Q Assisted                                   1346
Pre-Op Q Assisted By                                   0
Pre-Op Q Symptom Period                             1103
Pre-Op Q Previous Surgery                            915
Pre-Op Q Living Arrangements                        1680
Pre-Op Q Disability                                 7341
Heart Disease                                     114100
High Bp                                            77501
Stroke                                            123155
Circulation                                       118922
Lung Disease                                      114478
Diab

,Provider Code,Procedure,Revision Flag,Year,Age Band,Gender,Pre-Op Q Assisted,Pre-Op Q Assisted By,Pre-Op Q Symptom Period,Pre-Op Q Previous Surgery,Pre-Op Q Living Arrangements,Pre-Op Q Disability,Heart Disease,High Bp,Stroke,Circulation,Lung Disease,Diabetes,Kidney Disease,Nervous System,Liver Disease,Cancer,Depression,Arthritis,Pre-Op Q Mobility,Pre-Op Q Self-Care,Pre-Op Q Activity,Pre-Op Q Discomfort,Pre-Op Q Anxiety,Pre-Op Q EQ5D Index Profile,Pre-Op Q EQ5D Index,Post-Op Q Assisted,Post-Op Q Assisted By,Post-Op Q Living Arrangements,Post-Op Q Disability,Post-Op Q Mobility,Post-Op Q Self-Care,Post-Op Q Activity,Post-Op Q Discomfort,Post-Op Q Anxiety,Post-Op Q Satisfaction,Post-Op Q Sucess,Post-Op Q Allergy,Post-Op Q Bleeding,Post-Op Q Wound,Post-Op Q Urine,Post-Op Q Further Surgery,Post-Op Q Readmitted,Post-Op Q EQ5D Index Profile,Post-Op Q EQ5D Index,Hip Replacement EQ5D Index Post-Op Q Predicted,Pre-Op Q EQ VAS,Post-Op Q EQ VAS,Hip Replacement EQ VAS Post-Op Q Predicted,Hip Replacement Pre-Op Q Pain,Hip Replacement Pre-Op Q Sudden Pain,Hip Replacement Pre-Op Q Night Pain,Hip Replacement Pre-Op Q Washing,Hip Replacement Pre-Op Q Transport,Hip Replacement Pre-Op Q Dressing,Hip Replacement Pre-Op Q Shopping,Hip Replacement Pre-Op Q Walking,Hip Replacement Pre-Op Q Limping,Hip Replacement Pre-Op Q Stairs,Hip Replacement Pre-Op Q Standing,Hip Replacement Pre-Op Q Work,Hip Replacement Pre-Op Q Score,Hip Replacement Post-Op Q Pain,Hip Replacement Post-Op Q Sudden Pain,Hip Replacement Post-Op Q Night Pain,Hip Replacement Post-Op Q Washing,Hip Replacement Post-Op Q Transport,Hip Replacement Post-Op Q Dressing,Hip Replacement Post-Op Q Shopping,Hip Replacement Post-Op Q Walking,Hip Replacement Post-Op Q Limping,Hip Replacement Post-Op Q Stairs,Hip Replacement Post-Op Q Standing,Hip Replacement Post-Op Q Work,Hip Replacement Post-Op Q Score,Hip Replacement OHS Post-Op Q Predicted,CSVYear
124834,RYR,Hip Replacement,0,2018/19,80 to 89,2,2.0,0,2.0,2.0,2.0,2.0,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,2.0,1.0,3.0,3.0,1.0,21331,0.10,2.0,NaN,2.0,1.0,2.0,1.0,2.0,2.0,1.0,2.0,1.0,2.0,2.0,2.0,1.0,2.0,2.0,21221,0.69,0.78,85.0,60.0,8.269.750.656,0.0,3.0,1.0,2.0,2.0,3.0,2.0,2.0,0.0,2.0,1.0,0.0,18.0,3.0,3.0,3.0,3.0,3.0,3.0,2.0,3.0,3.0,3.0,4.0,4.0,37.0,3.906.359.677,2018/19
124835,RYR,Hip Replacement,0,2018/19,80 to 89,2,2.0,0,2.0,2.0,2.0,2.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,2.0,2.0,3.0,3.0,2.0,22332,-0.07,2.0,NaN,2.0,2.0,2.0,1.0,2.0,1.0,1.0,3.0,1.0,2.0,2.0,2.0,2.0,2.0,2.0,21211,0.81,0.79,31.0,49.0,7.572.149.595,0.0,4.0,0.0,2.0,2.0,3.0,1.0,4.0,0.0,1.0,2.0,1.0,20.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,4.0,3.0,3.0,4.0,4.0,46.0,4.169.309.415,2018/19
124836,RYR,Hip Replacement,0,2018/19,80 to 89,2,2.0,0,2.0,2.0,2.0,1.0,NaN,NaN,NaN,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,1.0,2.0,2.0,2.0,2.0,1.0,22221,0.59,2.0,NaN,2.0,NaN,2.0,1.0,1.0,1.0,1.0,1.0,1.0,2.0,2.0,2.0,2.0,2.0,2.0,21111,0.85,0.80,70.0,70.0,7.547.700.222,0.0,4.0,2.0,2.0,2.0,2.0,4.0,3.0,1.0,2.0,1.0,1.0,24.0,3.0,4.0,4.0,4.0,3.0,3.0,4.0,3.0,3.0,3.0,4.0,3.0,41.0,3.928.416.093,2018/19
124837,RYR,Hip Replacement,1,2018/19,80 to 89,2,1.0,0,2.0,1.0,1.0,2.0,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.0,1.0,2.0,2.0,1.0,21221,0.69,2.0,NaN,1.0,1.0,2.0,2.0,2.0,2.0,1.0,4.0,4.0,2.0,2.0,2.0,2.0,2.0,2.0,22221,0.59,0.81,100.0,85.0,8.710.635.693,1.0,2.0,1.0,2.0,2.0,2.0,0.0,2.0,0.0,0.0,2.0,0.0,14.0,1.0,1.0,3.0,2.0,2.0,2.0,1.0,2.0,1.0,2.0,3.0,1.0,21.0,3.361.814.753,2018/19
124838,RYR,Hip Replacement,0,2018/19,80 to 89,2,2.0,0,2.0,2.0,1.0,1.0,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2.0,1.0,2.0,2.0,1.0,21221,0.69,2.0,NaN,1.0,1.0,2.0,1.0,1.0,1.0,1.0,4.0,NaN,2.0,2.0,2.0,2.0,2.0,2.0,21111,0.85,0.81,75.0,60.0,764.610.392,0.0,2.0,1.0,2.0,1.0,1.0,1.0,1.0,0.0,1.0,1.0,1.0,12.0,4.0,4.0,4.0,4.0,3.0,4.0,2.0,4.0,0.0,2.0,4.0,4.0,39.0,3.570.394.365,2018/19
124839,RYR,Hip Replacement,1,2018/19,80 to 89,2,2.0,0,2.0,1.0,2.0,2.0,NaN,NaN,1.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.0,2.0,2.0,2.0,2.0,1.0,22221,0.59,2.0,NaN,2.

In [154]:
# Analyze missing values by Age_Band in df_selected
missing_by_age_selected = df.groupby('Age Band', dropna=False).agg(lambda x: x.isnull().sum())
print("Missing values by Age_Band in df_selected:")
print(missing_by_age_selected)

Missing values by Age_Band in df_selected:
           Provider Code  Procedure  Revision Flag  Year  Gender  \
Age Band                                                           
20 to 29               0          0              0     0       0   
30 to 39               0          0              0     0       0   
40 to 49               0          0              0     0       0   
50 to 59               0          0              0     0       0   
60 to 69               0          0              0     0       0   
70 to 79               0          0              0     0       0   
80 to 89               0          0              0     0       0   
90 to 120              0          0              0     0       0   
NaN                    0          0              0     0    9591   

           Pre-Op Q Assisted  Pre-Op Q Assisted By  Pre-Op Q Symptom Period  \
Age Band                                                                      
20 to 29                   0                     0

The following columns are defined in the dictionary with 1 = Yes and 9 = Missing. In these binary indicator fields, it is common data science practice to treat a missing value (9) as a "No" (0), as it implies the patient did not report the condition [Conversation History].
• Arthritis Indicator (ARTHRITIS)
• Cancer Indicator (CANCER)
• Circulation Indicator (CIRCULATION)
• Depression Indicator (DEPRESSION)
• Diabetes Indicator (DIABETES)
• Heart Disease Indicator (HEART_DISEASE)
• High Blood Pressure Indicator (HIGH_BP)
• Kidney Disease Indicator (KIDNEY_DISEASE)
• Liver Disease Indicator (LIVER_DISEASE)
• Lung Disease Indicator (LUNG_DISEASE)
• Nervous System Indicator (NERVOUS_SYSTEM)
• Stroke Indicator (STROKE)


In [155]:
#replace missing column values as described
indicator_columns = [
    'Arthritis',
    'Cancer',
    'Circulation',
    'Depression',
    'Diabetes',
    'Heart Disease',
    'High Bp',
    'Kidney Disease',
    'Liver Disease',
    'Lung Disease',
    'Nervous System',
    'Stroke'
]

# Replace NaN (originally 9) with 0 in these indicator columns
for col in indicator_columns:
    if col in df.columns:
        df[col] = df[col].fillna(0)
        print(f"Replaced missing values in {col} with 0.")
    else:
        print(f"Column {col} does not exist in the dataframe.")

print("All indicator columns updated.")

Replaced missing values in Arthritis with 0.
Replaced missing values in Cancer with 0.
Replaced missing values in Circulation with 0.
Replaced missing values in Depression with 0.
Replaced missing values in Diabetes with 0.
Replaced missing values in Heart Disease with 0.
Replaced missing values in High Bp with 0.
Replaced missing values in Kidney Disease with 0.
Replaced missing values in Liver Disease with 0.
Replaced missing values in Lung Disease with 0.
Replaced missing values in Nervous System with 0.
Replaced missing values in Stroke with 0.
All indicator columns updated.


Delete patient records where they havent responded for 12 questions related to oxford hop score 

these 12 columns represent the individual dimensions of the Oxford Hip Score (OHS), which is the primary disease-specific measure for hip replacements

We removed observations with missing values" to ensure that the models were trained on fully informative and structured data. By removing rows where these functional questions were skipped, you ensure your model follows a robust real-world benchmark

The sources identify that preoperative Q score dimensions are among the most important predictors for surgical success. Specifically, the question regarding limping (HR_Q1_LIMPING) is cited as a top predictor for postoperative outcomes. Imputing values for such high-impact variables would introduce significant noise and bias, potentially causing the model to misidentify "Non-Responders"

Hip Replacement Pre-Op Q Pain                      0.15
Hip Replacement Pre-Op Q Sudden Pain               0.89
Hip Replacement Pre-Op Q Night Pain                0.86
Hip Replacement Pre-Op Q Washing                   0.11
Hip Replacement Pre-Op Q Transport                 0.90
Hip Replacement Pre-Op Q Dressing                  0.90
Hip Replacement Pre-Op Q Shopping                  0.91
Hip Replacement Pre-Op Q Walking                   0.94
Hip Replacement Pre-Op Q Limping                   0.89
Hip Replacement Pre-Op Q Stairs                    0.93
Hip Replacement Pre-Op Q Standing                  0.91
Hip Replacement Pre-Op Q Work                      0.90

In [156]:
# Remove rows where any of the 12 Hip Replacement Pre-Op Q dimensions have missing values
hr_pre_cols = [col for col in df.columns if col.startswith('Hip Replacement Pre-Op Q') and col != 'Hip Replacement Pre-Op Q Score']

print("Hip Replacement Pre-Op Q columns (excluding Score):", hr_pre_cols)
print("Number of columns:", len(hr_pre_cols))

initial_shape = df.shape
df = df.dropna(subset=hr_pre_cols)
final_shape = df.shape

print(f"Removed {initial_shape[0] - final_shape[0]} rows with missing values in Hip Replacement Pre-Op Q dimensions.")
print(f"New dataframe shape: {final_shape}")

Hip Replacement Pre-Op Q columns (excluding Score): ['Hip Replacement Pre-Op Q Pain', 'Hip Replacement Pre-Op Q Sudden Pain', 'Hip Replacement Pre-Op Q Night Pain', 'Hip Replacement Pre-Op Q Washing', 'Hip Replacement Pre-Op Q Transport', 'Hip Replacement Pre-Op Q Dressing', 'Hip Replacement Pre-Op Q Shopping', 'Hip Replacement Pre-Op Q Walking', 'Hip Replacement Pre-Op Q Limping', 'Hip Replacement Pre-Op Q Stairs', 'Hip Replacement Pre-Op Q Standing', 'Hip Replacement Pre-Op Q Work']
Number of columns: 12
Removed 1417 rows with missing values in Hip Replacement Pre-Op Q dimensions.
New dataframe shape: (123427, 82)



Missing values for Hip Replacement Pre-Op Q Score t0_HR_Score and Hip Replacement Post-Op Q Score t1_HR_Score can be calaculated by adding up the 12 questions
Hip ReplacementScore = It is the sum total of the Oxford Hip Score (OHS), a validated 12-question survey used by the NHS to establish a baseline for patient health

Missing values for Pre-Op Q EQ5D Index Profile can be calculated by concatenating Pre-Op Q Mobility	Pre-Op Q Self-Care	Pre-Op Q Activity	Pre-Op Q Discomfort	Pre-Op Q Anxiety as string

Missing value for post-Op Q EQ5D Index Profile can be calcualted by concatinating Post-Op Q Mobility	Post-Op Q Self-Care	Post-Op Q Activity	Post-Op Q Discomfort	Post-Op Q Anxiety


In [157]:

# Calculate missing values for t0_HR_Score by summing the 12 t0_HR_ columns
t0_hr_cols = [col for col in df.columns if col.startswith('Hip Replacement Pre-Op Q') and col != 'Hip Replacement Pre-Op Q Score']

print("Hip Replacement Pre-Op Q columns (excluding Score):", t0_hr_cols)
print("Number of Hip Replacement Pre-Op Q columns:", len(t0_hr_cols))

# Ensure there are 12 columns
if len(t0_hr_cols) == 12:
    # Fill missing Hip Replacement Pre-Op Q Score with the sum of the 12 columns
    df['Hip Replacement Pre-Op Q Score'] = df['Hip Replacement Pre-Op Q Score'].fillna(df[t0_hr_cols].sum(axis=1))
    print("Missing values in Hip Replacement Pre-Op Q Score filled by summing the 12 Hip Replacement Pre-Op Q columns.")
else:
    print(f"Warning: Expected 12 Hip Replacement Pre-Op Q columns, but found {len(t0_hr_cols)}.")



Hip Replacement Pre-Op Q columns (excluding Score): ['Hip Replacement Pre-Op Q Pain', 'Hip Replacement Pre-Op Q Sudden Pain', 'Hip Replacement Pre-Op Q Night Pain', 'Hip Replacement Pre-Op Q Washing', 'Hip Replacement Pre-Op Q Transport', 'Hip Replacement Pre-Op Q Dressing', 'Hip Replacement Pre-Op Q Shopping', 'Hip Replacement Pre-Op Q Walking', 'Hip Replacement Pre-Op Q Limping', 'Hip Replacement Pre-Op Q Stairs', 'Hip Replacement Pre-Op Q Standing', 'Hip Replacement Pre-Op Q Work']
Number of Hip Replacement Pre-Op Q columns: 12
Missing values in Hip Replacement Pre-Op Q Score filled by summing the 12 Hip Replacement Pre-Op Q columns.


In [158]:
# Calculate missing values for t1_HR_Score by summing the 12 t1_HR_ columns
t1_hr_cols = [col for col in df.columns if col.startswith('Hip Replacement Post-Op Q') and col != 'Hip Replacement Post-Op Q Score' and 'Predicted' not in col]

print("Hip Replacement Post-Op Q columns (excluding Score and Predicted):", t1_hr_cols)
print("Number of Hip Replacement Post-Op Q columns:", len(t1_hr_cols))

# Ensure there are 12 columns
if len(t1_hr_cols) == 12:
    # Fill missing Hip Replacement Post-Op Q Score with the sum of the 12 columns
    df['Hip Replacement Post-Op Q Score'] = df['Hip Replacement Post-Op Q Score'].fillna(df[t1_hr_cols].sum(axis=1))
    print("Missing values in Hip Replacement Post-Op Q Score filled by summing the 12 Hip Replacement Post-Op Q columns.")
else:
    print(f"Warning: Expected 12 Hip Replacement Post-Op Q columns, but found {len(t1_hr_cols)}.")

Hip Replacement Post-Op Q columns (excluding Score and Predicted): ['Hip Replacement Post-Op Q Pain', 'Hip Replacement Post-Op Q Sudden Pain', 'Hip Replacement Post-Op Q Night Pain', 'Hip Replacement Post-Op Q Washing', 'Hip Replacement Post-Op Q Transport', 'Hip Replacement Post-Op Q Dressing', 'Hip Replacement Post-Op Q Shopping', 'Hip Replacement Post-Op Q Walking', 'Hip Replacement Post-Op Q Limping', 'Hip Replacement Post-Op Q Stairs', 'Hip Replacement Post-Op Q Standing', 'Hip Replacement Post-Op Q Work']
Number of Hip Replacement Post-Op Q columns: 12
Missing values in Hip Replacement Post-Op Q Score filled by summing the 12 Hip Replacement Post-Op Q columns.


 Remove Patient that has not responded to respective questions - 

Pre-Op Q Mobility                                  3.32
Pre-Op Q Self-Care                                 3.38
Pre-Op Q Activity                                  3.44
Pre-Op Q Discomfort                                4.16
Pre-Op Q Anxiety                                   3.88

Post-Op Q Mobility                                 2.11
Post-Op Q Self-Care                                1.92
Post-Op Q Activity                                 2.14
Post-Op Q Discomfort                               2.76
Post-Op Q Anxiety                                  2.18
 
 Loss of Clinical Precision: These dimensions are used to identify "Non-Responders". If you impute a "1" (No Problems) for a patient who actually had severe problems, you introduce noise that can lead the model to wrongly predict that a patient will not reach the Minimal Important Difference (MID).
• Impact on Calculated Index: The EQ-5D Index is calculated by subtracting specific weights from a 1.0 baseline based on these five digits. An imputed value would create a "hallucinated" index score, leading to selection bias.
• Invalid Value Errors: As established in our conversation history, you cannot replace these with 0. The valid range for these dimensions is 1 to 3 (1=No problems, 2=Moderate, 3=Extreme)

In [159]:
#remove patients that have not responded to above questions to avoid noise
eq5d_cols = [
    'Pre-Op Q Mobility', 'Pre-Op Q Self-Care', 'Pre-Op Q Activity', 'Pre-Op Q Discomfort', 'Pre-Op Q Anxiety',
    'Post-Op Q Mobility', 'Post-Op Q Self-Care', 'Post-Op Q Activity', 'Post-Op Q Discomfort', 'Post-Op Q Anxiety'
]

# Remove rows where any of these EQ-5D dimensions have null values
initial_shape = df.shape
df = df.dropna(subset=eq5d_cols)
final_shape = df.shape

print(f"Removed {initial_shape[0] - final_shape[0]} rows with missing values in EQ-5D dimensions.")
print(f"New dataframe shape: {final_shape}")

Removed 11608 rows with missing values in EQ-5D dimensions.
New dataframe shape: (111819, 82)


• Pre-Op Q Assisted (1.08% missing):
    ◦ Dictionary: 1 = Yes, 2 = No, 9 = Missing.
    ◦ Strategy: Since the missingness is tiny, perform Mode Imputation to "2" (No), which is the dominant value in the provided CSV excerpts.

In [160]:
# Mode imputation for Pre-Op Q Assisted
# Replace missing values (NaN) with 2 (No)
df.loc[:, 'Pre-Op Q Assisted'] = df['Pre-Op Q Assisted'].fillna(2)
print("Missing values in 'Pre-Op Q Assisted' filled with 2 (No).")

Missing values in 'Pre-Op Q Assisted' filled with 2 (No).


Pre-Op Q Symptom Period (0.88% missing):
    ◦ Dictionary: 1 = <1 yr, 2 = 1–5 yrs, 3 = 6–10 yrs, 4 = >10 yrs, 9 = Missing.
    ◦ Strategy: The sources identify Symptom Duration as a strong predictor for postoperative outcomes. Because this is a high-impact variable, do not use mode imputation. Listwise deletion for these specific rows is recommended to avoid introducing noise into a critical predictor

In [161]:
# Listwise deletion for Pre-Op Q Symptom Period
# Remove rows where 'Pre-Op Q Symptom Period' is missing
initial_shape = df.shape
df = df.dropna(subset=['Pre-Op Q Symptom Period'])
final_shape = df.shape

print(f"Removed {initial_shape[0] - final_shape[0]} rows with missing 'Pre-Op Q Symptom Period'.")
print(f"New dataframe shape: {final_shape}")

Removed 696 rows with missing 'Pre-Op Q Symptom Period'.
New dataframe shape: (111123, 82)


Pre-Op Q Previous Surgery (0.73% missing):
    ◦ Dictionary: 1 = Yes, 2 = No, 9 = Missing.
    ◦ Strategy: Given the extremely low missingness, Mode Imputation to "2" (No) is a safe way to preserve the other data points in those rows

In [162]:
# Mode imputation for Pre-Op Q Previous Surgery
# Replace missing values (NaN) with 2 (No)
df.loc[:, 'Pre-Op Q Previous Surgery'] = df['Pre-Op Q Previous Surgery'].fillna(2)
print("Missing values in 'Pre-Op Q Previous Surgery' filled with 2 (No).")

Missing values in 'Pre-Op Q Previous Surgery' filled with 2 (No).


Pre-Op Q Living Arrangements (1.35% missing):
    ◦ Dictionary: 1 = With family/partner, 2 = Alone, 3 = Nursing home, 4 = Other, 9 = Missing.
    ◦ Strategy: Keep "9" as a distinct category ("Unknown/Not Specified"). In predictive modelling, a patient's choice to skip a question about their home life can sometimes be an informative feature regarding their social support system

In [163]:
# Replace null with 9 to keep it as a separate category
df.loc[:, 'Pre-Op Q Living Arrangements'] = df['Pre-Op Q Living Arrangements'].fillna(9)
print("Missing values in 'Pre-Op Q Living Arrangements' filled with 9 (Unknown/Not Specified).")

Missing values in 'Pre-Op Q Living Arrangements' filled with 9 (Unknown/Not Specified).


Pre-Op Q Disability (5.88% missing):
Dictionary: 1 = Yes, 2 = No, 9 = Missing.
    ◦ Strategy: This is your highest missingness. Do not replace 9 with 0, as the dictionary defines "2" as the code for "No".
    ◦ Best approach: Treat "9" as a third categorical level ("Not Disclosed"). Patients may hesitate to self-identify as "disabled," and forcing them into "Yes" or "No" would introduce selection bias

In [164]:
#replace none with 9 to specify not disclosed category
df.loc[:, 'Pre-Op Q Disability'] = df['Pre-Op Q Disability'].fillna(9)
print("Missing values in 'Pre-Op Q Disability' filled with 9 (Not Disclosed).")

Missing values in 'Pre-Op Q Disability' filled with 9 (Not Disclosed).


In [165]:
# Function to calculate EQ-5D-5L index from profile using UK value set
def calculate_eq5d_index(profile):
    if pd.isna(profile) or len(str(profile)) != 5:
        return np.nan
    try:
        digits = [int(d) for d in str(profile)]
        if any(d not in [1,2,3] for d in digits):
            return np.nan  # Invalid or missing
    except:
        return np.nan
    
    # UK value set weights
    constant = 0.081
    n3_penalty = 0.269
    
    dim_weights = {
        0: {'2': 0.069, '3': 0.314},  # Mobility
        1: {'2': 0.104, '3': 0.214},  # Self-care
        2: {'2': 0.036, '3': 0.094},  # Usual activities
        3: {'2': 0.123, '3': 0.386},  # Pain/discomfort
        4: {'2': 0.071, '3': 0.236}   # Anxiety/depression
    }
    
    total_deduction = constant
    has_level_3 = False
    
    for i, level in enumerate(digits):
        if level == 2:
            total_deduction += dim_weights[i]['2']
        elif level == 3:
            total_deduction += dim_weights[i]['3']
            has_level_3 = True
    
    if has_level_3:
        total_deduction += n3_penalty
    
    index = 1.0 - total_deduction
    return index

# Verify the EQ-5D index calculation for existing values
df_verify = df.dropna(subset=['Pre-Op Q EQ5D Index']).copy()

df_verify['calculated_index'] = df_verify['Pre-Op Q EQ5D Index Profile'].apply(calculate_eq5d_index)

df_verify['difference'] = df_verify['Pre-Op Q EQ5D Index'] - df_verify['calculated_index']
print("Verification of EQ-5D index calculation:")
print(f"Number of non-missing Pre-Op Q EQ5D Index: {len(df_verify)}")
print(f"Mean difference: {df_verify['difference'].mean()}")
print(f"Max difference: {df_verify['difference'].max()}")
print(f"Min difference: {df_verify['difference'].min()}")
print(f"Number of exact matches (diff < 1e-6): {(df_verify['difference'].abs() < 1e-6).sum()}")

print("\nFirst 10 examples:")
print(df_verify[['Pre-Op Q EQ5D Index Profile', 'Pre-Op Q EQ5D Index', 'calculated_index', 'difference']].head(10))
# Fill missing Pre-Op Q EQ5D Index using Pre-Op Q EQ5D Index Profile

Verification of EQ-5D index calculation:
Number of non-missing Pre-Op Q EQ5D Index: 111123
Mean difference: 0.0003054183202397369
Max difference: 0.08099999999999996
Min difference: -1.6653345369377348e-16
Number of exact matches (diff < 1e-6): 110704

First 10 examples:
    Pre-Op Q EQ5D Index Profile  Pre-Op Q EQ5D Index  calculated_index  \
1                         22231             5.50e-02          5.50e-02   
2                         22332            -7.40e-02         -7.40e-02   
3                         22231             5.50e-02          5.50e-02   
4                         22232            -1.60e-02         -1.60e-02   
5                         21232             8.80e-02          8.80e-02   
6                         21321             3.64e-01          3.64e-01   
7                         22331            -3.00e-03         -3.00e-03   
8                         22221             5.87e-01          5.87e-01   
9                         11221             7.60e-01          

The verification output shows that the EQ-5D index calculation formula is correct for the vast majority of cases. Out of 117,824 non-missing t0_EQ5D_Index values:

117,354 (99.6%) are exact matches (difference < 1e-6, likely due to floating-point precision).
The mean difference is negligible (0.00032), and the maximum difference is 0.081, which could be due to rounding in the original data or slight variations in value sets.

In [166]:
# Fill missing t0_EQ5D_Index_Profile by concatenating the 5 EQ5D components
eq5d_cols = ['Pre-Op Q Mobility', 'Pre-Op Q Self-Care', 'Pre-Op Q Activity', 'Pre-Op Q Discomfort', 'Pre-Op Q Anxiety']

df['Pre-Op Q EQ5D Index Profile'] = df['Pre-Op Q EQ5D Index Profile'].fillna(
    df[eq5d_cols].astype(str).agg(''.join, axis=1)
)

print("Missing values in Pre-Op Q EQ5D Index Profile filled by concatenating the 5 EQ5D components.")
df['Pre-Op Q EQ5D Index'] = df['Pre-Op Q EQ5D Index'].fillna(
    df['Pre-Op Q EQ5D Index Profile'].apply(calculate_eq5d_index)
)

print("Missing values in Pre-Op Q EQ5D Index filled using the calculated index from Pre-Op Q EQ5D Index Profile.")

Missing values in Pre-Op Q EQ5D Index Profile filled by concatenating the 5 EQ5D components.
Missing values in Pre-Op Q EQ5D Index filled using the calculated index from Pre-Op Q EQ5D Index Profile.


In [167]:
# Fill missing t0_EQ5D_Index_Profile by concatenating the 5 EQ5D components
eq5d_cols = ['Post-Op Q Mobility', 'Post-Op Q Self-Care', 'Post-Op Q Activity', 'Post-Op Q Discomfort', 'Post-Op Q Anxiety']

df['Post-Op Q EQ5D Index Profile'] = df['Post-Op Q EQ5D Index Profile'].fillna(
    df[eq5d_cols].astype(str).agg(''.join, axis=1)
)

print("Missing values in Post-Op Q EQ5D Index Profile filled by concatenating the 5 EQ5D components.")
df['Post-Op Q EQ5D Index'] = df['Post-Op Q EQ5D Index'].fillna(
    df['Post-Op Q EQ5D Index Profile'].apply(calculate_eq5d_index)
)

print("Missing values in Post-Op Q EQ5D Index filled using the calculated index from Post-Op Q EQ5D Index Profile.")

Missing values in Post-Op Q EQ5D Index Profile filled by concatenating the 5 EQ5D components.
Missing values in Post-Op Q EQ5D Index filled using the calculated index from Post-Op Q EQ5D Index Profile.


Ignore columns -

post operation questions, the colmn with t1 and Q init
Hip Replacement EQ5D Index Post-Op Q Predicted - predicted by liner regression
Hip Replacement OHS Post-Op Q Predicted 
Hip Replacement EQ VAS Post-Op Q Predicted


In [168]:
# Identify columns to remove and keep
all_columns = df.columns.tolist()

# Columns to keep even if they start with 't1_'
keep_t1 = ['Post-Op Q EQ5D Index Profile', 'Post-Op Q EQ5D Index', 'Post-Op Q EQ VAS', 'Post-Op Q HR_Score']

# Remove columns starting with 't1_' (except the specified ones) and containing 'Predicted'
removed_columns = [col for col in all_columns if ('Post-Op' in col and col not in keep_t1) or 'Predicted' in col or col == 'CSVYear']

kept_columns = [col for col in all_columns if col not in removed_columns]

print("Columns to be removed:")
print(removed_columns)
print(f"\nTotal removed: {len(removed_columns)}")

print("\nColumns to be kept:")
print(kept_columns)
print(f"\nTotal kept: {len(kept_columns)}")

Columns to be removed:
['Post-Op Q Assisted', 'Post-Op Q Assisted By', 'Post-Op Q Living Arrangements', 'Post-Op Q Disability', 'Post-Op Q Mobility', 'Post-Op Q Self-Care', 'Post-Op Q Activity', 'Post-Op Q Discomfort', 'Post-Op Q Anxiety', 'Post-Op Q Satisfaction', 'Post-Op Q Sucess', 'Post-Op Q Allergy', 'Post-Op Q Bleeding', 'Post-Op Q Wound', 'Post-Op Q Urine', 'Post-Op Q Further Surgery', 'Post-Op Q Readmitted', 'Hip Replacement EQ5D Index Post-Op Q Predicted', 'Hip Replacement EQ VAS Post-Op Q Predicted', 'Hip Replacement Post-Op Q Pain', 'Hip Replacement Post-Op Q Sudden Pain', 'Hip Replacement Post-Op Q Night Pain', 'Hip Replacement Post-Op Q Washing', 'Hip Replacement Post-Op Q Transport', 'Hip Replacement Post-Op Q Dressing', 'Hip Replacement Post-Op Q Shopping', 'Hip Replacement Post-Op Q Walking', 'Hip Replacement Post-Op Q Limping', 'Hip Replacement Post-Op Q Stairs', 'Hip Replacement Post-Op Q Standing', 'Hip Replacement Post-Op Q Work', 'Hip Replacement Post-Op Q Score', 

In [169]:
# Create new dataframe with selected columns
df_selected = df[kept_columns].copy()

print("New dataframe created with selected columns.")
print("Shape of new dataframe:", df_selected.shape)
print("Columns in new dataframe:")
print(df_selected.columns.tolist())

New dataframe created with selected columns.
Shape of new dataframe: (111123, 48)
Columns in new dataframe:
['Provider Code', 'Procedure', 'Revision Flag', 'Year', 'Age Band', 'Gender', 'Pre-Op Q Assisted', 'Pre-Op Q Assisted By', 'Pre-Op Q Symptom Period', 'Pre-Op Q Previous Surgery', 'Pre-Op Q Living Arrangements', 'Pre-Op Q Disability', 'Heart Disease', 'High Bp', 'Stroke', 'Circulation', 'Lung Disease', 'Diabetes', 'Kidney Disease', 'Nervous System', 'Liver Disease', 'Cancer', 'Depression', 'Arthritis', 'Pre-Op Q Mobility', 'Pre-Op Q Self-Care', 'Pre-Op Q Activity', 'Pre-Op Q Discomfort', 'Pre-Op Q Anxiety', 'Pre-Op Q EQ5D Index Profile', 'Pre-Op Q EQ5D Index', 'Post-Op Q EQ5D Index Profile', 'Post-Op Q EQ5D Index', 'Pre-Op Q EQ VAS', 'Post-Op Q EQ VAS', 'Hip Replacement Pre-Op Q Pain', 'Hip Replacement Pre-Op Q Sudden Pain', 'Hip Replacement Pre-Op Q Night Pain', 'Hip Replacement Pre-Op Q Washing', 'Hip Replacement Pre-Op Q Transport', 'Hip Replacement Pre-Op Q Dressing', 'Hip Rep

In [170]:
# List columns with missing values in df_selected
missing_cols = df_selected.columns[df_selected.isnull().any()].tolist()

print("Columns with missing values in df_selected:")
print(missing_cols)
print(f"\nTotal columns with missing values: {len(missing_cols)}")

# Also show the count of missing values per column
print("\nMissing values count per column:")
print(df_selected.isnull().sum()[df_selected.isnull().sum() > 0])

Columns with missing values in df_selected:
['Age Band', 'Gender', 'Pre-Op Q EQ VAS', 'Post-Op Q EQ VAS']

Total columns with missing values: 4

Missing values count per column:
Age Band            8686
Gender              8686
Pre-Op Q EQ VAS     6372
Post-Op Q EQ VAS    3435
dtype: int64


In [171]:
# Save the cleaned dataframe to a Parquet file
# Convert object columns to string for parquet compatibility
df_selected = df_selected.astype({col: 'string' for col in df_selected.select_dtypes(include='object').columns})

df_selected.to_parquet('cleaned_hip_replacement_data.parquet', engine='fastparquet')
print("Dataframe saved to 'cleaned_hip_replacement_data.parquet'")

Dataframe saved to 'cleaned_hip_replacement_data.parquet'


for age band and Gender =

 Listwise Deletion (Removal)
The most robust approach for your machine learning model is to remove these 8,686 observations entirely.
• Methodological Alignment: In the research paper evaluating hip replacements, the researchers explicitly stated: "We removed observations with missing values" to ensure the classifiers were trained on fully informative data.
• Statistical Impact: Deleting 8,686 rows from a total population of 130,945 observations represents a loss of only approximately 6.6% of your data. This is a minor trade-off to ensure your model doesn't attempt to learn patterns from "masked" demographics.
• Avoiding Noise: Since age and gender are important casemix-adjustment variables used by the NHS to predict outcomes, attempting to impute them would introduce significant noise